File: app/src/main/java/fr/free/nrw/commons/auth/LoginActivity.java
package fr.free.nrw.commons.auth;

import static android.view.KeyEvent.KEYCODE_ENTER;
import static android.view.View.VISIBLE;
import static android.view.inputmethod.EditorInfo.IME_ACTION_DONE;
import static fr.free.nrw.commons.CommonsApplication.loginMessageIntentKey;
import static fr.free.nrw.commons.CommonsApplication.loginUsernameIntentKey;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.KeyEvent;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.TextView;
import androidx.annotation.ColorRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.NavUtils;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProvider;
import fr.free.nrw.commons.BuildConfig;
import fr.free.nrw.commons.R;
import fr.free.nrw.commons.Utils;
import fr.free.nrw.commons.auth.login.LoginCallback;
import fr.free.nrw.commons.auth.login.LoginResult;
import fr.free.nrw.commons.contributions.MainActivity;
import fr.free.nrw.commons.databinding.ActivityLoginBinding;
import fr.free.nrw.commons.di.ApplicationlessInjection;
import fr.free.nrw.commons.di.CommonsApplicationComponent;
import fr.free.nrw.commons.kvstore.JsonKvStore;
import fr.free.nrw.commons.utils.ActivityUtils;
import fr.free.nrw.commons.utils.ConfigUtils;
import fr.free.nrw.commons.utils.SystemThemeUtils;
import fr.free.nrw.commons.utils.ViewUtil;
import java.util.Locale;
import java.util.Objects;
import javax.inject.Inject;
import javax.inject.Named;
import timber.log.Timber;

public class LoginActivity extends AppCompatActivity {

    @Inject
    SessionManager sessionManager;

    @Inject
    @Named("default_preferences")
    JsonKvStore applicationKvStore;

    @Inject
    SystemThemeUtils systemThemeUtils;

    private LoginViewModel model;
    private ActivityLoginBinding binding;
    private ProgressDialog progressDialog;
    private LoginTextWatcher textWatcher = new LoginTextWatcher();
    private LoginCallback loginCallback = new LoginCallbackImpl();

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        CommonsApplicationComponent commonsApplicationComponent = ApplicationlessInjection
            .getInstance(this.getApplicationContext())
            .getCommonsApplicationComponent();
        commonsApplicationComponent.inject(this);

        model = new ViewModelProvider(this,
            new LoginViewModelFactory(commonsApplicationComponent)).get(LoginViewModel.class);

        boolean isDarkTheme = systemThemeUtils.isDeviceInNightMode();
        setTheme(isDarkTheme ? R.style.DarkAppTheme : R.style.LightAppTheme);

        binding = ActivityLoginBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        binding.loginUsername.addTextChangedListener(textWatcher);
        binding.loginPassword.addTextChangedListener(textWatcher);
        binding.loginTwoFactor.addTextChangedListener(textWatcher);

        binding.skipLogin.setOnClickListener(view -> skipLogin());
        binding.forgotPassword.setOnClickListener(view -> Utils.handleWebUrl(
            this, Uri.parse(BuildConfig.FORGOT_PASSWORD_URL)));
        binding.aboutPrivacyPolicy.setOnClickListener(view -> Utils.handleWebUrl(
            this, Uri.parse(BuildConfig.PRIVACY_POLICY_URL)));
        binding.signUpButton.setOnClickListener(view -> startActivity(new Intent(this, SignupActivity.class)));
        binding.loginButton.setOnClickListener(view -> performLogin());

        binding.loginPassword.setOnEditorActionListener(this::onEditorAction);
        binding.loginPassword.setOnFocusChangeListener(this::onPasswordFocusChanged);

        if (ConfigUtils.isBetaFlavour()) {
            binding.loginCredentials.setText(getString(R.string.login_credential));
        } else {
            binding.loginCredentials.setVisibility(View.GONE);
        }

        String message = getIntent().getStringExtra(loginMessageIntentKey);
        if (message != null) {
            showMessage(message, R.color.secondaryDarkColor);
        }

        String username = getIntent().getStringExtra(loginUsernameIntentKey);
        if (username != null) {
            binding.loginUsername.setText(username);
        }
    }

    /**
     * Hides the keyboard if the user's focus is not on the password (hasFocus is false).
     *
     * @param view     The keyboard
     * @param hasFocus Set to true if the keyboard has focus
     */
    void onPasswordFocusChanged(View view, boolean hasFocus) {
        if (!hasFocus) {
            ViewUtil.hideKeyboard(view);
        }
    }

    boolean onEditorAction(TextView textView, int actionId, KeyEvent keyEvent) {
        if (binding.loginButton.isEnabled()) {
            if (actionId == IME_ACTION_DONE) {
                performLogin();
                return true;
            } else if ((keyEvent != null) && keyEvent.getKeyCode() == KEYCODE_ENTER) {
                performLogin();
                return true;
            }
        }
        return false;
    }

    protected void skipLogin() {
        new AlertDialog.Builder(this).setTitle(R.string.skip_login_title)
            .setMessage(R.string.skip_login_message)
            .setCancelable(false)
            .setPositiveButton(R.string.yes, (dialog, which) -> {
                dialog.cancel();
                performSkipLogin();
            })
            .setNegativeButton(R.string.no, (dialog, which) -> dialog.cancel())
            .show();
    }

    @Override
    protected void onResume() {
        super.onResume();

        if (sessionManager.getCurrentAccount() != null
            && sessionManager.isUserLoggedIn()) {
            applicationKvStore.putBoolean("login_skipped", false);
            startMainActivity();
        }

        if (applicationKvStore.getBoolean("login_skipped", false)) {
            performSkipLogin();
        }

    }

    @Override
    protected void onDestroy() {
        try {
            // To prevent leaked window when finish() is called, see http://stackoverflow.com/questions/32065854/activity-has-leaked-window-at-alertdialog-show-method
            if (progressDialog != null && progressDialog.isShowing()) {
                hideProgress();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        binding.loginUsername.removeTextChangedListener(textWatcher);
        binding.loginPassword.removeTextChangedListener(textWatcher);
        binding.loginTwoFactor.removeTextChangedListener(textWatcher);

        binding = null;
        super.onDestroy();
    }

    public void performLogin() {
        final String username = Objects.requireNonNull(binding.loginUsername.getText()).toString();
        final String password = Objects.requireNonNull(binding.loginPassword.getText()).toString();
        final String twoFactorCode = Objects.requireNonNull(binding.loginTwoFactor.getText())
            .toString();

        showLoggingProgressBar();

        model.doLogin(
            username, password, twoFactorCode, Locale.getDefault().getLanguage(), loginCallback
        );
    }

    private void hideProgress() {
        runOnUiThread(() -> {
            if (progressDialog != null) {
                progressDialog.dismiss();
                progressDialog = null;
            }
        });
    }

    private void showPasswordResetPrompt() {
        showMessageAndCancelDialog(getString(R.string.you_must_reset_your_passsword));
    }

    /**
     * This function is called when user skips the login. It redirects the user to Explore
     * Activity.
     */
    private void performSkipLogin() {
        applicationKvStore.putBoolean("login_skipped", true);
        MainActivity.startYourself(this);
        finish();
    }

    private void showLoggingProgressBar() {
        progressDialog = new ProgressDialog(this);
        progressDialog.setIndeterminate(true);
        progressDialog.setTitle(getString(R.string.logging_in_title));
        progressDialog.setMessage(getString(R.string.logging_in_message));
        progressDialog.setCanceledOnTouchOutside(false);
        progressDialog.show();
    }

    private void onLoginSuccess(LoginResult loginResult) {
        sessionManager.setUserLoggedIn(true);
        sessionManager.updateAccount(loginResult);
        showMessage(R.string.login_success, R.color.primaryDarkColor);
        hideProgress();
        startMainActivity();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                NavUtils.navigateUpFromSameTask(this);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    public void askUserForTwoFactorAuth() {
        hideProgress();
        binding.twoFactorContainer.setVisibility(VISIBLE);
        binding.loginTwoFactor.setVisibility(VISIBLE);
        binding.loginTwoFactor.requestFocus();
        InputMethodManager imm = (InputMethodManager) getSystemService(
            Context.INPUT_METHOD_SERVICE);
        imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, InputMethodManager.HIDE_IMPLICIT_ONLY);
        showMessageAndCancelDialog(R.string.login_failed_2fa_needed);
    }

    public void showMessageAndCancelDialog(@StringRes int resId) {
        showMessage(resId, R.color.secondaryDarkColor);
        if (progressDialog != null) {
            progressDialog.cancel();
        }
    }

    public void showMessageAndCancelDialog(String error) {
        showMessage(error, R.color.secondaryDarkColor);
        if (progressDialog != null) {
            progressDialog.cancel();
        }
    }

    public void startMainActivity() {
        ActivityUtils.startActivityWithFlags(this, MainActivity.class,
            Intent.FLAG_ACTIVITY_SINGLE_TOP);
        finish();
    }

    private void showMessage(@StringRes int resId, @ColorRes int colorResId) {
        binding.errorMessage.setText(getString(resId));
        binding.errorMessage.setTextColor(ContextCompat.getColor(this, colorResId));
        binding.errorMessageContainer.setVisibility(VISIBLE);
    }

    private void showMessage(String message, @ColorRes int colorResId) {
        binding.errorMessage.setText(message);
        binding.errorMessage.setTextColor(ContextCompat.getColor(this, colorResId));
        binding.errorMessageContainer.setVisibility(VISIBLE);
    }

    private class LoginTextWatcher implements TextWatcher {

        @Override
        public void beforeTextChanged(CharSequence charSequence, int start, int count, int after) {
        }

        @Override
        public void onTextChanged(CharSequence charSequence, int start, int count, int after) {
        }

        @Override
        public void afterTextChanged(Editable editable) {
            boolean enabled = binding.loginUsername.getText().length() != 0 &&
                binding.loginPassword.getText().length() != 0 &&
                (BuildConfig.DEBUG || binding.loginTwoFactor.getText().length() != 0 ||
                    binding.loginTwoFactor.getVisibility() != VISIBLE);
            binding.loginButton.setEnabled(enabled);
        }
    }

    public static void startYourself(Context context) {
        Intent intent = new Intent(context, LoginActivity.class);
        context.startActivity(intent);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        // if progressDialog is visible during the configuration change  then store state as  true else false so that
        // we maintain visibility of progressDailog after configuration change
        outState.putBoolean("ProgressDailog_state",
            progressDialog != null && progressDialog.isShowing());
        outState.putString("errorMessage",
            binding.errorMessage.getText().toString()); //Save the errorMessage
        outState.putString("username",
            binding.loginUsername.getText().toString()); // Save the username
        outState.putString("password",
            binding.loginPassword.getText().toString()); // Save the password
    }

    @Override
    protected void onRestoreInstanceState(final Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        binding.loginUsername.setText(savedInstanceState.getString("username"));
        binding.loginPassword.setText(savedInstanceState.getString("password"));
        if (savedInstanceState.getBoolean("ProgressDailog_state")) {
            performLogin();
        }
        String errorMessage = savedInstanceState.getString("errorMessage");
        if (sessionManager.isUserLoggedIn()) {
            showMessage(R.string.login_success, R.color.primaryDarkColor);
        } else {
            showMessage(errorMessage, R.color.secondaryDarkColor);
        }
    }

    private class LoginCallbackImpl implements LoginCallback {

        @Override
        public void success(@NonNull LoginResult loginResult) {
            Log.e("###", "success = Thread = "+Thread.currentThread().getName());
            runOnUiThread(() -> {
                Timber.d("Login Success");
                hideProgress();
                onLoginSuccess(loginResult);
            });
        }

        @Override
        public void twoFactorPrompt(@NonNull Throwable caught, @Nullable String token) {
            Log.e("###", "twoFactorPrompt = Thread = "+Thread.currentThread().getName());
            runOnUiThread(() -> {
                Timber.d("Requesting 2FA prompt");
                hideProgress();
                askUserForTwoFactorAuth();
            });
        }

        @Override
        public void passwordResetPrompt(@Nullable String token) {
            Log.e("###", "passwordResetPrompt = Thread = "+Thread.currentThread().getName());
            runOnUiThread(() -> {
                Timber.d("Showing password reset prompt");
                hideProgress();
                showPasswordResetPrompt();
            });
        }

        @Override
        public void error(@NonNull Throwable caught) {
            Log.e("###", "error = Thread = "+Thread.currentThread().getName());
            runOnUiThread(() -> {
                Timber.e(caught);
                hideProgress();
                showMessageAndCancelDialog(caught.getLocalizedMessage());
            });
        }
    }
}

File: app/src/main/java/fr/free/nrw/commons/auth/LoginActivity.kt
package fr.free.nrw.commons.auth

import android.accounts.AccountAuthenticatorActivity
import android.app.ProgressDialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.KeyEvent
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.TextView
import androidx.annotation.ColorRes
import androidx.annotation.StringRes
import androidx.annotation.VisibleForTesting
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.app.NavUtils
import androidx.core.content.ContextCompat
import androidx.core.view.WindowCompat
import fr.free.nrw.commons.BuildConfig
import fr.free.nrw.commons.CommonsApplication
import fr.free.nrw.commons.R
import fr.free.nrw.commons.auth.login.LoginCallback
import fr.free.nrw.commons.auth.login.LoginClient
import fr.free.nrw.commons.auth.login.LoginResult
import fr.free.nrw.commons.contributions.MainActivity
import fr.free.nrw.commons.databinding.ActivityLoginBinding
import fr.free.nrw.commons.di.ApplicationlessInjection
import fr.free.nrw.commons.kvstore.JsonKvStore
import fr.free.nrw.commons.utils.applyEdgeToEdgeAllInsets
import fr.free.nrw.commons.utils.AbstractTextWatcher
import fr.free.nrw.commons.utils.ActivityUtils.startActivityWithFlags
import fr.free.nrw.commons.utils.ConfigUtils.isBetaFlavour
import fr.free.nrw.commons.utils.SystemThemeUtils
import fr.free.nrw.commons.utils.ViewUtil.hideKeyboard
import fr.free.nrw.commons.utils.handleKeyboardInsets
import fr.free.nrw.commons.utils.handleWebUrl
import io.reactivex.disposables.CompositeDisposable
import timber.log.Timber
import java.util.Locale
import javax.inject.Inject
import javax.inject.Named

class LoginActivity : AccountAuthenticatorActivity() {
    @Inject
    lateinit var sessionManager: SessionManager

    @Inject
    @field:Named("default_preferences")
    lateinit var applicationKvStore: JsonKvStore

    @Inject
    lateinit var loginClient: LoginClient

    @Inject
    lateinit var systemThemeUtils: SystemThemeUtils

    private var binding: ActivityLoginBinding? = null
    private var progressDialog: ProgressDialog? = null
    private val textWatcher = AbstractTextWatcher(::onTextChanged)
    private val compositeDisposable = CompositeDisposable()
    private val delegate: AppCompatDelegate by lazy {
        AppCompatDelegate.create(this, null)
    }
    private var lastLoginResult: LoginResult? = null

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ApplicationlessInjection
            .getInstance(this.applicationContext)
            .commonsApplicationComponent
            .inject(this)

        val isDarkTheme = systemThemeUtils.isDeviceInNightMode()
        setTheme(if (isDarkTheme) R.style.DarkAppTheme else R.style.LightAppTheme)
        delegate.installViewFactory()
        delegate.onCreate(savedInstanceState)

        WindowCompat.getInsetsController(window, window.decorView)
            .isAppearanceLightStatusBars = !isDarkTheme

        WindowCompat.setDecorFitsSystemWindows(window, false)

        binding = ActivityLoginBinding.inflate(layoutInflater)
        applyEdgeToEdgeAllInsets(binding!!.root)
        binding!!.root.handleKeyboardInsets()
        with(binding!!) {
            setContentView(root)

            loginUsername.addTextChangedListener(textWatcher)
            loginPassword.addTextChangedListener(textWatcher)
            loginTwoFactor.addTextChangedListener(textWatcher)

            skipLogin.setOnClickListener { skipLogin() }
            forgotPassword.setOnClickListener { forgotPassword() }
            aboutPrivacyPolicy.setOnClickListener { onPrivacyPolicyClicked() }
            signUpButton.setOnClickListener { signUp() }
            loginButton.setOnClickListener { performLogin() }
            loginPassword.setOnEditorActionListener { textView, actionId, keyEvent ->
                if (binding!!.loginButton.isEnabled && isTriggerAction(actionId, keyEvent)) {
                    if (actionId == EditorInfo.IME_ACTION_NEXT && lastLoginResult != null) {
                        askUserForTwoFactorAuthWithKeyboard()
                        true
                    } else {
                        performLogin()
                        true
                    }
                } else {
                    false
                }
            }

            loginPassword.onFocusChangeListener =
                View.OnFocusChangeListener(::onPasswordFocusChanged)

            if (isBetaFlavour) {
                loginCredentials.text = getString(R.string.login_credential)
            } else {
                loginCredentials.visibility = View.GONE
            }

            intent.getStringExtra(CommonsApplication.LOGIN_MESSAGE_INTENT_KEY)?.let {
                showMessage(it, R.color.secondaryDarkColor)
            }

            intent.getStringExtra(CommonsApplication.LOGIN_USERNAME_INTENT_KEY)?.let {
                loginUsername.setText(it)
            }
        }
    }

    @VisibleForTesting
    fun askUserForTwoFactorAuthWithKeyboard() {
        if (binding == null) {
            Timber.w("Binding is null, reinitializing in askUserForTwoFactorAuthWithKeyboard")
            binding = ActivityLoginBinding.inflate(layoutInflater)
            setContentView(binding!!.root)
        }
        progressDialog!!.dismiss()
        if (binding != null) {
            with(binding!!) {
                twoFactorContainer.visibility = View.VISIBLE
                twoFactorContainer.hint = getString(if (lastLoginResult is LoginResult.EmailAuthResult) R.string.email_auth_code else R.string._2fa_code)
                loginTwoFactor.visibility = View.VISIBLE
                loginTwoFactor.requestFocus()

                val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
                imm.showSoftInput(loginTwoFactor, InputMethodManager.SHOW_IMPLICIT)

                loginTwoFactor.setOnEditorActionListener { _, actionId, event ->
                    if (actionId == EditorInfo.IME_ACTION_DONE ||
                        (event != null && event.keyCode == KeyEvent.KEYCODE_ENTER && event.action == KeyEvent.ACTION_DOWN)) {
                        performLogin()
                        true
                    } else {
                        false
                    }
                }
            }
        } else {
            Timber.e("Binding is null in askUserForTwoFactorAuthWithKeyboard after reinitialization attempt")
        }
        showMessageAndCancelDialog(getString(if (lastLoginResult is LoginResult.EmailAuthResult) R.string.login_failed_email_auth_needed else R.string.login_failed_2fa_needed))
    }
    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        delegate.onPostCreate(savedInstanceState)
    }

    override fun onResume() {
        super.onResume()

        if (sessionManager.currentAccount != null && sessionManager.isUserLoggedIn) {
            applicationKvStore.putBoolean("login_skipped", false)
            startMainActivity()
        }

        if (applicationKvStore.getBoolean("login_skipped", false)) {
            performSkipLogin()
        }
    }

    override fun onDestroy() {
        compositeDisposable.clear()
        try {
            // To prevent leaked window when finish() is called, see http://stackoverflow.com/questions/32065854/activity-has-leaked-window-at-alertdialog-show-method
            if (progressDialog?.isShowing == true) {
                progressDialog!!.dismiss()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        with(binding!!) {
            loginUsername.removeTextChangedListener(textWatcher)
            loginPassword.removeTextChangedListener(textWatcher)
            loginTwoFactor.removeTextChangedListener(textWatcher)
        }
        delegate.onDestroy()
        loginClient.cancel()
        binding = null
        super.onDestroy()
    }

    override fun onStart() {
        super.onStart()
        delegate.onStart()
    }

    override fun onStop() {
        super.onStop()
        delegate.onStop()
    }

    override fun onPostResume() {
        super.onPostResume()
        delegate.onPostResume()
    }

    override fun setContentView(view: View, params: ViewGroup.LayoutParams) {
        delegate.setContentView(view, params)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                NavUtils.navigateUpFromSameTask(this)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        // if progressDialog is visible during the configuration change  then store state as  true else false so that
        // we maintain visibility of progressDialog after configuration change
        if (progressDialog != null && progressDialog!!.isShowing) {
            outState.putBoolean(SAVE_PROGRESS_DIALOG, true)
        } else {
            outState.putBoolean(SAVE_PROGRESS_DIALOG, false)
        }
        outState.putString(
            SAVE_ERROR_MESSAGE,
            binding!!.errorMessage.text.toString()
        ) //Save the errorMessage
        outState.putString(
            SAVE_USERNAME,
            binding!!.loginUsername.text.toString()
        ) // Save the username
        outState.putString(
            SAVE_PASSWORD,
            binding!!.loginPassword.text.toString()
        ) // Save the password
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        binding!!.loginUsername.setText(savedInstanceState.getString(SAVE_USERNAME))
        binding!!.loginPassword.setText(savedInstanceState.getString(SAVE_PASSWORD))
        if (savedInstanceState.getBoolean(SAVE_PROGRESS_DIALOG)) {
            performLogin()
        }
        val errorMessage = savedInstanceState.getString(SAVE_ERROR_MESSAGE)
        if (sessionManager.isUserLoggedIn) {
            showMessage(R.string.login_success, R.color.primaryDarkColor)
        } else {
            showMessage(errorMessage, R.color.secondaryDarkColor)
        }
    }

    /**
     * Hides the keyboard if the user's focus is not on the password (hasFocus is false).
     * @param view The keyboard
     * @param hasFocus Set to true if the keyboard has focus
     */
    private fun onPasswordFocusChanged(view: View, hasFocus: Boolean) {
        if (!hasFocus) {
            hideKeyboard(view)
        }
    }

    private fun onEditorAction(textView: TextView, actionId: Int, keyEvent: KeyEvent?) =
        if (binding!!.loginButton.isEnabled && isTriggerAction(actionId, keyEvent)) {
            performLogin()
            true
        } else false

    private fun isTriggerAction(actionId: Int, keyEvent: KeyEvent?) =
        actionId == EditorInfo.IME_ACTION_NEXT || actionId == EditorInfo.IME_ACTION_DONE || keyEvent?.keyCode == KeyEvent.KEYCODE_ENTER

    private fun skipLogin() {
        AlertDialog.Builder(this)
            .setTitle(R.string.skip_login_title)
            .setMessage(R.string.skip_login_message)
            .setCancelable(false)
            .setPositiveButton(R.string.yes) { dialog: DialogInterface, which: Int ->
                dialog.cancel()
                performSkipLogin()
            }
            .setNegativeButton(R.string.no) { dialog: DialogInterface, which: Int ->
                dialog.cancel()
            }
            .show()
    }

    private fun forgotPassword() =
        handleWebUrl(this, Uri.parse(BuildConfig.FORGOT_PASSWORD_URL))

    private fun onPrivacyPolicyClicked() =
        handleWebUrl(this, Uri.parse(BuildConfig.PRIVACY_POLICY_URL))

    private fun signUp() =
        startActivity(Intent(this, SignupActivity::class.java))

    @VisibleForTesting
    fun performLogin() {
        Timber.d("Login to start!")
        val username = binding!!.loginUsername.text.toString()
        val password = binding!!.loginPassword.text.toString()
        val twoFactorCode = binding!!.loginTwoFactor.text.toString()

        showLoggingProgressBar()
        loginClient.doLogin(username,
            password,
            lastLoginResult,
            twoFactorCode,
            Locale.getDefault().language,
            object : LoginCallback {
                override fun success(loginResult: LoginResult) = runOnUiThread {
                    Timber.d("Login Success")
                    progressDialog!!.dismiss()
                    onLoginSuccess(loginResult)
                }

                override fun twoFactorPrompt(loginResult: LoginResult, caught: Throwable, token: String?) = runOnUiThread {
                    Timber.d("Requesting 2FA prompt")
                    progressDialog!!.dismiss()
                    lastLoginResult = loginResult
                    askUserForTwoFactorAuthWithKeyboard()
                }

                override fun emailAuthPrompt(loginResult: LoginResult, caught: Throwable, token: String?) = runOnUiThread {
                    Timber.d("Requesting email auth prompt")
                    progressDialog!!.dismiss()
                    lastLoginResult = loginResult
                    askUserForTwoFactorAuthWithKeyboard()
                }

                override fun passwordResetPrompt(token: String?) = runOnUiThread {
                    Timber.d("Showing password reset prompt")
                    progressDialog!!.dismiss()
                    showPasswordResetPrompt()
                }

                override fun error(caught: Throwable) = runOnUiThread {
                    Timber.e(caught)
                    progressDialog!!.dismiss()
                    showMessageAndCancelDialog(caught.localizedMessage ?: "")
                }
            }
        )
    }

    private fun showPasswordResetPrompt() =
        showMessageAndCancelDialog(getString(R.string.you_must_reset_your_passsword))

    /**
     * This function is called when user skips the login.
     * It redirects the user to Explore Activity.
     */
    private fun performSkipLogin() {
        applicationKvStore.putBoolean("login_skipped", true)
        MainActivity.startYourself(this)
        finish()
    }

    private fun showLoggingProgressBar() {
        progressDialog = ProgressDialog(this).apply {
            isIndeterminate = true
            setTitle(getString(R.string.logging_in_title))
            setMessage(getString(R.string.logging_in_message))
            setCancelable(false)
        }
        progressDialog!!.show()
    }

    private fun onLoginSuccess(loginResult: LoginResult) {
        compositeDisposable.clear()
        sessionManager.setUserLoggedIn(true)
        sessionManager.updateAccount(loginResult)
        progressDialog!!.dismiss()
        showSuccessAndDismissDialog()
        startMainActivity()
    }

    override fun getMenuInflater(): MenuInflater =
        delegate.menuInflater

    @VisibleForTesting
    fun askUserForTwoFactorAuth() {
        if (binding == null) {
            Timber.w("Binding is null, reinitializing in askUserForTwoFactorAuth")
            binding = ActivityLoginBinding.inflate(layoutInflater)
            setContentView(binding!!.root)
        }
        progressDialog!!.dismiss()
        if (binding != null) {
            with(binding!!) {
                twoFactorContainer.visibility = View.VISIBLE
                twoFactorContainer.hint = getString(if (lastLoginResult is LoginResult.EmailAuthResult) R.string.email_auth_code else R.string._2fa_code)
                loginTwoFactor.visibility = View.VISIBLE
                loginTwoFactor.requestFocus()

                loginTwoFactor.setOnEditorActionListener { _, actionId, event ->
                    if (actionId == EditorInfo.IME_ACTION_DONE ||
                        (event != null && event.keyCode == KeyEvent.KEYCODE_ENTER && event.action == KeyEvent.ACTION_DOWN)) {
                        performLogin()
                        true
                    } else {
                        false
                    }
                }
            }
        } else {
            Timber.e("Binding is null in askUserForTwoFactorAuth after reinitialization attempt")
        }
        val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
        imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, InputMethodManager.HIDE_IMPLICIT_ONLY)
        showMessageAndCancelDialog(getString(if (lastLoginResult is LoginResult.EmailAuthResult) R.string.login_failed_email_auth_needed else R.string.login_failed_2fa_needed))
    }

    @VisibleForTesting
    fun showMessageAndCancelDialog(@StringRes resId: Int) {
        showMessage(resId, R.color.secondaryDarkColor)
        progressDialog?.cancel()
    }

    @VisibleForTesting
    fun showMessageAndCancelDialog(error: String) {
        showMessage(error, R.color.secondaryDarkColor)
        progressDialog?.cancel()
    }

    @VisibleForTesting
    fun showSuccessAndDismissDialog() {
        showMessage(R.string.login_success, R.color.primaryDarkColor)
        progressDialog!!.dismiss()
    }

    @VisibleForTesting
    fun startMainActivity() {
        startActivityWithFlags(this, MainActivity::class.java, Intent.FLAG_ACTIVITY_SINGLE_TOP)
        finish()
    }

    private fun showMessage(@StringRes resId: Int, @ColorRes colorResId: Int) = with(binding!!) {
        errorMessage.text = getString(resId)
        errorMessage.setTextColor(ContextCompat.getColor(this@LoginActivity, colorResId))
        errorMessageContainer.visibility = View.VISIBLE
    }

    private fun showMessage(message: String?, @ColorRes colorResId: Int) = with(binding!!) {
        errorMessage.text = message
        errorMessage.setTextColor(ContextCompat.getColor(this@LoginActivity, colorResId))
        errorMessageContainer.visibility = View.VISIBLE
    }

    private fun onTextChanged(text: String) {
        val enabled =
            binding!!.loginUsername.text!!.length != 0 && binding!!.loginPassword.text!!.length != 0 &&
                    (BuildConfig.DEBUG || binding!!.loginTwoFactor.text!!.length != 0 || binding!!.loginTwoFactor.visibility != View.VISIBLE)
        binding!!.loginButton.isEnabled = enabled
    }

    companion object {
        fun startYourself(context: Context) =
            context.startActivity(Intent(context, LoginActivity::class.java))

        const val SAVE_PROGRESS_DIALOG: String = "ProgressDialog_state"
        const val SAVE_ERROR_MESSAGE: String = "errorMessage"
        const val SAVE_USERNAME: String = "username"
        const val SAVE_PASSWORD: String = "password"
    }
}
